wkcuber/api/properties.py

```diff
@@ -0,0 +1,220 @@
+    "float64": "double",
+}
+
+
+def _snake_to_camel_case(snake_case_name: str) -> str:
+    parts = snake_case_name.split("_")
+    return parts[0] + "".join(part.title() for part in parts[1:])
+
+
+# --- View configuration --------------------
+
+
+@attr.define
+class DatasetViewConfiguration:
+    """
+    Stores information on how the dataset is shown in webknossos by default.
+    """
+
+    four_bit: Optional[bool] = None
+    interpolation: Optional[bool] = None
+    render_missing_data_black: Optional[bool] = None
+    loading_strategy: Optional[str] = None
+    segmentation_pattern_opacity: Optional[int] = None
+    zoom: Optional[float] = None
+    position: Optional[Tuple[int, int, int]] = None
+    rotation: Optional[Tuple[int, int, int]] = None
+
+
+@attr.define
+class LayerViewConfiguration:
+    """
+    Stores information on how the dataset is shown in webknossos by default.
+    """
+
+    color: Optional[Tuple[int, int, int]] = None
+    alpha: Optional[float] = None
+    intensity_range: Optional[Tuple[float, float]] = None
+    min: Optional[float] = None  # pylint: disable=redefined-builtin
+    max: Optional[float] = None  # pylint: disable=redefined-builtin
+    is_disabled: Optional[bool] = None
+    is_inverted: Optional[bool] = None
+    is_in_edit_mode: Optional[bool] = None
+
+
+# --- Property --------------------
+
+
+@attr.define
+class MagViewProperties:
+    resolution: Union[int, Mag]
+    cube_length: int
+
+
+@attr.define
+class LayerProperties:
+    name: str
+    category: str
+    bounding_box: BoundingBox
+    element_class: str
+    wkw_resolutions: List[MagViewProperties]
+    data_format: str
+    num_channels: Optional[int] = None
+    default_view_configuration: Optional[LayerViewConfiguration] = None
+
+
+@attr.define
+class SegmentationLayerProperties(LayerProperties):
+    largest_segment_id: int = -1
+    mappings: List[str] = []
+
+
+@attr.define
+class DatasetProperties:
+    id: Dict[str, str]
+    scale: Tuple[float, float, float]
+    data_layers: List[Union[SegmentationLayerProperties, LayerProperties]]
+    default_view_configuration: Optional[DatasetViewConfiguration] = None
+
+
+# --- Converter --------------------
+
+dataset_converter = cattr.Converter()
+
+# register (un-)structure hooks for non-attr-classes
+bbox_to_wkw: Callable[[BoundingBox], dict] = lambda o: o.as_wkw()
+dataset_converter.register_unstructure_hook(BoundingBox, bbox_to_wkw)
+dataset_converter.register_structure_hook(
+    BoundingBox, lambda d, _: BoundingBox.from_wkw(d)
+)
+
+mag_to_array: Callable[[Mag], List[int]] = lambda o: o.to_array()
+dataset_converter.register_unstructure_hook(Mag, mag_to_array)
+dataset_converter.register_structure_hook(Mag, lambda d, _: Mag(d))
+
+# register (un-)structure hooks for attr-classes
+for cls in [
+    DatasetProperties,
+    LayerProperties,
+    SegmentationLayerProperties,
+    MagViewProperties,
+    DatasetViewConfiguration,
+    LayerViewConfiguration,
+]:
```

Comment:
Instead of this loop, you could register (un-)structuring factories:
```
def snake_to_camel_case_overrides(attrs_type):
    return {
        a.name: override(omit_if_default=True, rename=_snake_to_camel_case(a.name))
        for a in attr.fields(attrs_type)
    }

register_unstructure_hook_factory(attrs.has, lambda attrs_type: make_dict_unstructure_fn(cls, dataset_converter, **snake_to_camel_case_overrides(attrs_type)))
# same for structuring
```

Reply:
I agree that is looks cleaner, but `cattr.Converter.register_unstructure_hook_factory` seems to be added in `1.9.0` which is not released yet.